File: library/src/main/java/com/anthonycr/bonsai/Schedulers.java
/*
 * Copyright (C) 2016 Anthony C. Restaino
 * <p/>
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 * <p/>
 * http://www.apache.org/licenses/LICENSE-2.0
 * <p/>
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
package com.anthonycr.bonsai;

import android.os.Looper;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import java.util.concurrent.Executor;
import java.util.concurrent.Executors;

/**
 * A class of default {@link Scheduler} provided
 * for use with {@link Stream}, {@link Single},
 * and {@link Completable}.
 * <p>
 * If the options available here are not sufficient,
 * implement {@link Scheduler} and create your own.
 */
@SuppressWarnings("WeakerAccess")
public final class Schedulers {

    @Nullable private static Scheduler mainScheduler;
    @Nullable private static Scheduler workerScheduler;
    @Nullable private static Scheduler ioScheduler;

    private Schedulers() {
        throw new UnsupportedOperationException("This class is not instantiable");
    }

    /**
     * A worker scheduler. Backed by a fixed
     * thread pool containing 4 thread.
     */
    private static class WorkerScheduler implements Scheduler {

        private final Executor worker = Executors.newFixedThreadPool(4);

        @Override
        public void execute(@NonNull Runnable command) {
            worker.execute(command);
        }
    }

    /**
     * A single threaded scheduler. Backed by a
     * single thread in an executor.
     */
    private static class SingleThreadedScheduler implements Scheduler {

        private final Executor singleThreadExecutor = Executors.newSingleThreadExecutor();

        @Override
        public void execute(@NonNull Runnable command) {
            singleThreadExecutor.execute(command);
        }
    }

    /**
     * A scheduler backed by an executor.
     */
    private static class ExecutorScheduler implements Scheduler {

        @NonNull
        private final Executor backingExecutor;

        public ExecutorScheduler(@NonNull Executor executor) {
            backingExecutor = executor;
        }

        @Override
        public void execute(@NonNull Runnable command) {
            backingExecutor.execute(command);
        }
    }

    /**
     * Creates a scheduler from an executor instance.
     *
     * @param executor the executor to use to create
     *                 the Scheduler.
     * @return a valid Scheduler backed by an executor.
     */
    @NonNull
    public static Scheduler from(@NonNull Executor executor) {
        return new ExecutorScheduler(executor);
    }

    /**
     * A scheduler that points to the
     * current thread. Useful when you
     * are not on the main thread and
     * need to observe on that thread.
     *
     * @return a scheduler associated with
     * the current thread.
     */
    @NonNull
    public static Scheduler current() {
        if (Looper.myLooper() == null) {
            Looper.prepare();
        }

        // Assert that the looper is not null
        // since we just prepared it if it was.
        Looper looper = Looper.myLooper();
        Preconditions.checkNonNull(looper);

        return new ThreadScheduler(looper);
    }

    /**
     * Creates a new Scheduler that
     * creates a new thread and does
     * all work on it.
     *
     * @return a scheduler associated
     * with a new single thread.
     */
    @NonNull
    public static Scheduler newSingleThreadedScheduler() {
        return new SingleThreadedScheduler();
    }

    /**
     * The worker thread Scheduler, will
     * execute work on any one of multiple
     * threads.
     *
     * @return a non-null Scheduler.
     */
    @NonNull
    public static Scheduler worker() {
        if (workerScheduler == null) {
            workerScheduler = new WorkerScheduler();
        }
        return workerScheduler;
    }

    /**
     * The main thread. All work will
     * be done on the single main thread.
     *
     * @return a non-null Scheduler that does work on the main thread.
     */
    @NonNull
    public static Scheduler main() {
        if (mainScheduler == null) {
            mainScheduler = new ThreadScheduler(Looper.getMainLooper());
        }
        return mainScheduler;
    }

    /**
     * The io scheduler. All work will be
     * done on a single thread.
     *
     * @return a non-null Scheduler that does
     * work on a single thread off the main thread.
     */
    @NonNull
    public static Scheduler io() {
        if (ioScheduler == null) {
            ioScheduler = new SingleThreadedScheduler();
        }
        return ioScheduler;
    }
}

File: library/src/main/java/com/anthonycr/bonsai/Schedulers.kt
/*
 * Copyright (C) 2016 Anthony C. Restaino
 * <p/>
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 * <p/>
 * http://www.apache.org/licenses/LICENSE-2.0
 * <p/>
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
package com.anthonycr.bonsai

import android.os.Handler
import android.os.Looper

import java.util.concurrent.Executor
import java.util.concurrent.Executors

/**
 * A class of default [Scheduler] provided
 * for use with [Stream], [Single],
 * and [Completable].
 *
 *
 * If the options available here are not sufficient,
 * implement [Scheduler] and create your own.
 */
object Schedulers {

    /**
     * A worker scheduler. Backed by a fixed
     * thread pool containing 4 thread.
     */
    private class WorkerScheduler : Scheduler {

        private val worker = Executors.newFixedThreadPool(4)

        override fun execute(runnable: () -> Unit) {
            worker.execute(runnable)
        }

    }

    /**
     * A single threaded scheduler. Backed by a
     * single thread in an executor.
     */
    private class SingleThreadedScheduler : Scheduler {

        private val singleThreadExecutor = Executors.newSingleThreadExecutor()

        override fun execute(runnable: () -> Unit) {
            singleThreadExecutor.execute(runnable)
        }

    }

    /**
     * A scheduler backed by an executor.
     */
    private class ExecutorScheduler(private val backingExecutor: Executor) : Scheduler {

        override fun execute(runnable: () -> Unit) {
            backingExecutor.execute(runnable)
        }

    }

    /**
     * A scheduler that executes tasks immediately.
     */
    private class ImmediateScheduler : Scheduler {

        override fun execute(runnable: () -> Unit) {
            runnable()
        }

    }

    private val mainScheduler: Scheduler by lazy { ThreadScheduler(Looper.getMainLooper()) }
    private val workerScheduler: Scheduler by lazy { WorkerScheduler() }
    private val ioScheduler: Scheduler by lazy { SingleThreadedScheduler() }
    private val immediateScheduler: Scheduler by lazy { ImmediateScheduler() }

    /**
     * Creates a scheduler from an executor instance.
     *
     * @param executor the executor to use to create
     * the Scheduler.
     * @return a valid Scheduler backed by an executor.
     */
    @JvmStatic
    fun from(executor: Executor): Scheduler = ExecutorScheduler(executor)

    /**
     * Creates a scheduler from a handler. The scheduler
     * will post tasks to the looper associated with the
     * handler.
     *
     * @param handler the handler used to create the Scheduler.
     * @return a valid Scheduler backed by a handler.
     */
    @JvmStatic
    fun from(handler: Handler): Scheduler = ThreadScheduler(handler.looper)

    /**
     * A scheduler that executes tasks synchronously
     * on the calling thread. If you use this scheduler
     * as the subscribe-on scheduler, then the work will
     * execute synchronously on the current thread. If
     * you use this scheduler as the observe-on thread,
     * then you will receive events on whatever thread
     * was used as that subscribe-on thread.
     *
     * @return a synchronous scheduler.
     */
    @JvmStatic
    fun immediate(): Scheduler = immediateScheduler

    /**
     * Creates a new Scheduler that
     * creates a new thread and does
     * all work on it.
     *
     * @return a scheduler associated
     * with a new single thread.
     */
    @JvmStatic
    fun newSingleThreadedScheduler(): Scheduler = SingleThreadedScheduler()

    /**
     * The worker thread Scheduler, will
     * execute work on any one of multiple
     * threads.
     *
     * @return a non-null Scheduler.
     */
    @JvmStatic
    fun worker(): Scheduler = workerScheduler

    /**
     * The main thread. All work will
     * be done on the single main thread.
     *
     * @return a non-null Scheduler that does work on the main thread.
     */
    @JvmStatic
    fun main(): Scheduler = mainScheduler

    /**
     * The io scheduler. All work will be
     * done on a single thread.
     *
     * @return a non-null Scheduler that does
     * work on a single thread off the main thread.
     */
    @JvmStatic
    fun io(): Scheduler = ioScheduler
}
